llvm/lib/Target/AIE/AIE2Subtarget.cpp

```diff
@@ -76,3 +76,53 @@ const RegisterBankInfo *AIE2Subtarget::getRegBankInfo() const {
 InstructionSelector *AIE2Subtarget::getInstructionSelector() const {
   return InstSelector.get();
 }
+
+std::bitset<32>
+AIE2Subtarget::getMemoryBanksFromAddressSpace(unsigned AddrSpace) const {
+  using namespace AIE2;
+  std::bitset<32> MemoryBanks;
+
+  switch (static_cast<AddressSpaces>(AddrSpace)) {
+  case AddressSpaces::a:
+    MemoryBanks.set(static_cast<int>(DMBanks::A));
+    break;
+  case AddressSpaces::b:
+    MemoryBanks.set(static_cast<int>(DMBanks::B));
+    break;
+  case AddressSpaces::c:
+    MemoryBanks.set(static_cast<int>(DMBanks::C));
+    break;
+  case AddressSpaces::d:
+    MemoryBanks.set(static_cast<int>(DMBanks::D));
+    break;
+  case AddressSpaces::ab:
+    MemoryBanks.set(static_cast<int>(DMBanks::A))
+        .set(static_cast<int>(DMBanks::B));
+    break;
+  case AddressSpaces::ac:
+    MemoryBanks.set(static_cast<int>(DMBanks::A))
+        .set(static_cast<int>(DMBanks::C));
+    break;
+  case AddressSpaces::ad:
+    MemoryBanks.set(static_cast<int>(DMBanks::A))
+        .set(static_cast<int>(DMBanks::D));
+    break;
+  case AddressSpaces::bc:
+    MemoryBanks.set(static_cast<int>(DMBanks::B))
+        .set(static_cast<int>(DMBanks::C));
+    break;
+  case AddressSpaces::bd:
+    MemoryBanks.set(static_cast<int>(DMBanks::B))
+        .set(static_cast<int>(DMBanks::D));
+    break;
+  case AddressSpaces::cd:
+    MemoryBanks.set(static_cast<int>(DMBanks::C))
+        .set(static_cast<int>(DMBanks::D));
+    break;
+  default:
+    // For unimplemented cases assume none
```

Comment:
I think this is inconsistent. DMBanks::none should conflict with every other bank

Reply:
I think this is inconsistent. DMBanks::none should conflict with every other bank
```
--> There are two way of looking at how to handle DMBanks::none 
Case 1. If we consider it as it conflicts with every other bank then due to lack of info we end up forcing 2 vector loads from different bank in two different bundle leaving HW no choice.
Case 2. If we consider it as conflicting with none and like previous case if the 2 vector loads are in different banks then HW will send them and no conflict but if they do conflict HW will add a stall and handle it.
Given these two cases and HW support to handle conflict for case of DMBanks::none I would prefer we let it not conflict with other
```

I would like to be complete, so not allow unimplemented cases here.
`The only unimplemented case here is DMBanks::stack, currently we cannot provide exact information of which bank the stack belongs so and is also not annotated by us or the user. Also user might use addresspace(AS) by themself and have some AS which is outside the scope.`

Note, that A/B/C/D/AB/.... are virtual banks and have nothing to do with the physical allocation.

Reply:
I would vouch for a consistent and conservative approach for now, i.e.

- AddressSpaces::a conflicts with `DMBanks::A`
- AddressSpaces::ac conflicts with both `DMBanks::A` and `DMBanks::C`
- Default conflicts with every bank
- etc.

If that's too conservative, we can adjust in a follow-up PR.

Reply:
Ok sounds good to me!

Reply:
On second thought, I do not think the idea of `Default conflicts with every bank` aligns with the AIE-ML arch. If we assume this was the case then at first place the HW would not have to support Core-Stall on bank conflict

Reply:
If we do track conflicts between loads and stores to the same bank, then yes, this is probably too conservative to assume unknown banks will conflict. That worked nicely if we only cared about loads, as we would prevent using LDA/LDB in parallel when not sure. Preventing LD/ST in parallel is a bigger problem.

Reply:
I think the bottom line is we never bothered about memory banks, since the hardware would correct it. That means the only rationale to do this is to be more precise, and give the user handles to tell what the situation is. If we say that unannotated banks don't interfere with others, we are undermining our own rationale. It would be different if 'no annotation' would be distinct from an 'assume no conflict' annotation.